Don't add newline at the start of triple-quotes

dra27 opened the pull request:
This follows on from https://github.com/ocaml/opam/issues/3995, https://github.com/ocaml/opam-repository/pull/14910 and https://github.com/ocaml/opam-repository/pull/14912

`OpamPrinter.format_value` adds a newline after a `"""`, but this is not removed by the lexer. I think this has gone largely unnoticed (I'd spotted that synopsis + description was adding a blank line, but not looked into it much further).

This is one fix - the other would be to have the first newline after `"""` ignored. However, the whole escaping based on `\n` seems weird - the purpose of `"""` is to have strings where `"` doesn't need escaping?

rjbou commented:
The purpose of `"""` is to not have to escape inside it, but the information is lost once _lexed_, single & triple quote are `string`. But on printing, all simple string are simple quoted (and everything escaped in it), otherwise (in case of newline), triple quoted. I think that the origin of this behaviour is that usually triple quote are used fort `description:` that is used to be composed by several paragraphs.
ping @altgr

Blaisorblade commented:
After seeing this issue, ocaml/opam-repository#15371 and ocaml/opam-repository#15496, I wonder why this sort of issue is possible.

It *seems* that when comparing package definitions with upstream ones, opam normalizes them in some way (that involves pretty-printing — I assume parsing and pretty-printing?), but apparently only definitions of installed packages, not the ones from upstream.

Blaisorblade commented:
Also, regardless of my speculations, does any of you understand why `conf-gnome-icon-theme3` has spurious upstream changes and if https://github.com/ocaml/opam-repository/pull/15496 is a sensible workaround/fix? It seems to work, but I struggle to believe it.

MSoegtropIMC commented:
I would appreciate if this PR gets merged. It burns a lot of CPU time that packages are rebuilt needlessly cause of string spacing issues. And it takes a lot of human time to find out the root cause.

Of cause it would be even better if opam would normalise upstream opam files in the same way it normalises saved opam files before comparing. Then it wouldn't matter in which way files are pretty printed. I think there are other issues in comparing upstream with stored opam files like changing single item lists to items.

rjbou commented:
After discussion with @AltGr, the best fix is to ignore the newline in the lexer

kit-ty-kate commented:
any news on this? Some users are still hitting this issue

dra27 commented:
> I think that the origin of this behaviour is that usually triple quote are used fort `description:` that is used to be composed by several paragraphs.

That's what it's used for, but in terms of the actual format, `"` handles in the same way? The only difference is not having to escape `"` inside `"""`, right? Everything else is identical between both string types in the lexer isn't it? Newlines are preserved and backslashes must be escaped in both?

dra27 commented:
More for the record than anything, but isn't altering the lexer a higher risk change, given the number of other programs which use the lexer, relative to the pretty printer?

AltGr commented:
Hm, indeed I thought that `"""` was expected to behave that way but I can't find a trace, so I probably got confused...
Isn't the issue simply that the printer should write
```
  str: """\
str contents"""
```
rather than
```
  str: """
str contents"""
```
?

dra27 commented:
Yeah, that should work, I think? Possibly with one adaptation - if the string actually begins with `\n` then there's no need for the escape! So in other words the pretty printer guarantees the string begins on the newline and also correctly preserves whether it actually began with `\n`?

Blaisorblade commented:
-then Format.fprintf fmt "\"\"\"\n%s\"\"\""
+then Format.fprintf fmt "\"\"\"\\\n%s\"\"\""

Blaisorblade commented:
@AltGr That also sounds good (and hopefully, equivalent to the current change). Can that be done by
```diff
-then Format.fprintf fmt "\"\"\"\n%s\"\"\""
+then Format.fprintf fmt "\"\"\"\\\n%s\"\"\""
```

> Possibly with one adaptation - if the string actually begins with `\n` then there's no need for the escape! So in other words the pretty printer guarantees the string begins on the newline and also correctly preserves whether it actually began with `\n`?

Just a user, but that might warrant more tests; is the extra conditional necessary, beyond giving a nicer output? But if that doesn't take more time to get done, sure.

dra27 committed:
Escape any newline at the start of triple-quotes

The OpamPrinter.format_value formats a string which contains newlines
as:

"""
line 1
line 2"""

The problem before is that the newline after the opening triple quote is
not escaped, so the string represented gains a newline. It should in
fact be:

"""\
line 1
line 2"""

so that round-tripping works. This commit adds a further optimisation so
that strings which actually begin with a newline display as:

"""
line 1
line 2"""

dra27 commented:
Rebased and tested with the most recent conf-dbm problem. Note that the _first_ time you run opam after this opam will still want to do the upgrade (in order to re-normalise the file).

I've tested the various cases of this too - the commit message contains the full description.

Blaisorblade commented:
FWIW, the code LGTM; my last comment was about actual testcases in the testsuite, but since no testsuite is present, that would be a separate question, not worth holding this up.

rjbou commented:
Thanks!
